File: zoom_bot.py
import asyncio
import random
import time
import sys
from playwright.async_api import async_playwright

# ============================================================
# CONFIGURATION
# ============================================================

JOIN_URL = "https://bytexl-in.zoom.us/w/87508297509?tk=Jwab1SSwtXE_pvbtNcHafwa2tqnfgysJfPZP7pS8Cz8.DQkAAAAUX-anJRZYb09MRWd2WFRDU0FwY2ZLY1ZKT2ZBAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA&pwd=wloDbtKaR91Ui2VqKxn9yqIm92pw8c.1"

# Bot 1
BOT1_FIRST = "Rahul"
BOT1_LAST = "Sharma"
BOT1_EMAIL = f"rahul.sharma{random.randint(100,999)}@gmail.com"

# Bot 2
BOT2_FIRST = "Priya"
BOT2_LAST = "Patel"
BOT2_EMAIL = f"priya.patel{random.randint(100,999)}@gmail.com"

DURATION_SECONDS = 3600  # 1 hour

# ============================================================

STEALTH_SCRIPT = """
Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
try { if (!window.chrome) { Object.defineProperty(window, 'chrome', { get: () => ({ runtime: {}, loadTimes: function() {}, csi: function() {}, app: {} }) }); } } catch(e) {}
Object.defineProperty(navigator, 'plugins', { get: () => [1,2,3,4,5] });
Object.defineProperty(navigator, 'languages', { get: () => ['en-US', 'en'] });
"""

def jitter(base, variance=0.3):
    return base + random.uniform(-variance * base, variance * base)


async def click_text(page, texts, timeout=10):
    """
    Wait for an element containing any of the given texts, then click it.
    Returns True if clicked, False if not found within timeout seconds.
    """
    for text in texts:
        try:
            # Try button
            btn = await page.wait_for_selector(f"button:has-text('{text}')", timeout=timeout * 1000)
            if btn:
                await btn.click()
                print(f"  ✅ Clicked: '{text}'")
                return True
        except:
            pass
        try:
            # Try anchor/div/span
            el = await page.wait_for_selector(f"a:has-text('{text}'), div[role='button']:has-text('{text}'), span:has-text('{text}')", timeout=timeout * 1000)
            if el:
                await el.click()
                print(f"  ✅ Clicked: '{text}'")
                return True
        except:
            pass
    return False


async def click_button_evaluate(page, text_parts):
    """Click using JavaScript evaluate for more reliability."""
    parts_json = str(text_parts)
    result = await page.evaluate(f"""() => {{
        const parts = {parts_json};
        const allEls = document.querySelectorAll('button, a, div[role="button"], span[role="button"], input[type="submit"], input[type="button"]');
        for (const el of allEls) {{
            const t = (el.textContent || el.value || '').toLowerCase().trim();
            for (const p of parts) {{
                if (t.includes(p.toLowerCase())) {{
                    el.click();
                    return 'clicked';
                }}
            }}
        }}
        return 'not-found';
    }}""")
    if result == 'clicked':
        print(f"  ✅ Clicked via JS: {text_parts[0]}...")
        return True
    return False


async def fill_form_fields(page, values):
    """Fill visible text inputs in order."""
    filled = 0
    inputs = await page.query_selector_all("input:not([type='hidden']):not([type='checkbox']):not([type='radio']):not([type='submit']):not([type='button'])")
    for inp in inputs:
        if filled >= len(values):
            break
        try:
            current = await inp.input_value()
            if not current:
                await inp.click()
                await asyncio.sleep(random.uniform(0.1, 0.3))
                await inp.fill(values[filled], delay=random.randint(40, 120))
                print(f"  ✅ Filled: {values[filled]}")
                filled += 1
        except:
            pass
    
    # Fallback for remaining fields
    if filled < len(values):
        remaining = values[filled:]
        for selector, val in [
            ("input[placeholder*='irst'], input[aria-label*='irst'], input[name='firstName']", remaining[0] if len(remaining) > 0 else None),
            ("input[placeholder*='ast'], input[aria-label*='ast'], input[name='lastName']", remaining[1] if len(remaining) > 1 else None),
            ("input[type='email'], input[placeholder*='email'], input[aria-label*='email'], input[name='email']", remaining[2] if len(remaining) > 2 else None),
        ]:
            if val:
                try:
                    el = await page.query_selector(selector)
                    if el:
                        curr = await el.input_value()
                        if not curr:
                            await el.click()
                            await asyncio.sleep(0.2)
                            await el.fill(val, delay=random.randint(40, 120))
                            print(f"  ✅ Filled (fallback): {val}")
                except:
                    pass
    return filled


async def run_single_bot(first_name, last_name, email, display_name, duration):
    print(f"\n{'='*50}")
    print(f"STARTING BOT: {display_name} ({email})")
    print(f"{'='*50}")

    async with async_playwright() as p:
        browser = await p.chromium.launch(
            headless=True,
            args=["--no-sandbox", "--disable-setuid-sandbox", "--disable-dev-shm-usage", "--disable-gpu", "--mute-audio"],
        )

        context = await browser.new_context(
            viewport={"width": 1920, "height": 1080},
            locale="en-US",
            timezone_id="Asia/Kolkata",
            user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
        )

        await context.add_init_script(STEALTH_SCRIPT)
        page = await context.new_page()

        # =============================================================
        # OPEN LINK
        # =============================================================
        print(f"[1] Opening Zoom link...")
        await page.goto(JOIN_URL, wait_until="domcontentloaded", timeout=60000)
        await asyncio.sleep(jitter(5.0))
        await page.screenshot(path=f"01_{first_name}.png")

        # Handle cookie popup if present
        await click_text(page, ["Accept all cookies", "Accept All Cookies", "Accept all", "Accept"], timeout=3)
        await asyncio.sleep(jitter(2.0))

        # =============================================================
        # STEP 1: Fill form (First Name, Last Name, Email)
        # =============================================================
        print(f"[2] Filling form: First Name, Last Name, Email...")
        await asyncio.sleep(jitter(2.0))
        await fill_form_fields(page, [first_name, last_name, email])

        # =============================================================
        # STEP 2: Click "Register and Join"
        # =============================================================
        print(f"[3] Looking for 'Register and Join'...")
        await asyncio.sleep(jitter(1.0))
        for attempt in range(10):
            found = await click_button_evaluate(page, ["register and join", "register", "join meeting", "submit", "join now"])
            if found:
                break
            await asyncio.sleep(jitter(1.5))
        await asyncio.sleep(jitter(4.0))

        # =============================================================
        # STEP 3: Click "Continue without audio and video"
        # =============================================================
        print(f"[4] Looking for 'Continue without audio and video'...")
        for attempt in range(10):
            found = await click_button_evaluate(page, [
                "continue without audio and video",
                "continue without",
                "continue",
                "skip",
                "no thanks",
                "don't connect audio"
            ])
            if found:
                break
            await asyncio.sleep(jitter(1.5))
        await asyncio.sleep(jitter(3.0))

        # =============================================================
        # STEP 4: Click "Join from Browser"
        # =============================================================
        print(f"[5] Looking for 'Join from Browser'...")
        
        # Try text-based click first
        browser_found = False
        for attempt in range(10):
            found = await click_button_evaluate(page, [
                "join from your browser",
                "join from browser",
                "from your browser",
                "browser"
            ])
            if found:
                browser_found = True
                break
            # Also try clicking any link containing "browser"
            try:
                link = await page.query_selector("a:has-text('browser'), a:has-text('Browser')")
                if link:
                    await link.click()
                    print(f"  ✅ Clicked browser link (direct)")
                    browser_found = True
                    break
            except:
                pass
            await asyncio.sleep(jitter(1.5))
        
        if not browser_found:
            # Fallback: go to WC URL directly
            try:
                wc_url = "https://bytexl-in.zoom.us/wc/join/87508297509"
                await page.goto(wc_url, wait_until="domcontentloaded", timeout=30000)
                await asyncio.sleep(jitter(4.0))
                print(f"  Used direct WC URL (fallback)")
            except:
                pass
        
        await asyncio.sleep(jitter(3.0))

        # =============================================================
        # STEP 5: Click "Continue without microphone and camera" (1st)
        # =============================================================
        print(f"[6] First 'Continue without microphone and camera'...")
        for attempt in range(10):
            found = await click_button_evaluate(page, [
                "continue without microphone and camera",
                "continue without microphone",
                "continue without camera",
                "continue without",
                "continue",
                "skip",
                "no thanks",
                "don't allow",
                "block"
            ])
            if found:
                break
            await asyncio.sleep(jitter(1.5))
        await asyncio.sleep(jitter(3.0))

        # =============================================================
        # STEP 6: Click "Continue without microphone and camera" (2nd)
        # =============================================================
        print(f"[7] Second 'Continue without microphone and camera'...")
        for attempt in range(10):
            found = await click_button_evaluate(page, [
                "continue without microphone and camera",
                "continue without microphone",
                "continue without camera",
                "continue without",
                "continue",
                "skip",
                "no thanks",
                "don't allow",
                "block"
            ])
            if found:
                break
            await asyncio.sleep(jitter(1.5))
        await asyncio.sleep(jitter(3.0))

        # =============================================================
        # STEP 7: Click "Join"
        # =============================================================
        print(f"[8] Looking for 'Join'...")
        for attempt in range(10):
            found = await click_button_evaluate(page, [
                "join",
                "join meeting",
                "join now",
                "enter meeting",
                "join the meeting",
                "enter"
            ])
            if found:
                break
            await asyncio.sleep(jitter(1.5))
        
        await asyncio.sleep(jitter(5.0))

        # Handle "Join with Computer Audio" inside meeting
        print(f"[9] Checking for audio prompt...")
        await asyncio.sleep(jitter(3.0))
        await click_button_evaluate(page, [
            "join with computer audio", "join audio", "listen only",
            "use computer audio", "join audio by computer"
        ])
        await asyncio.sleep(jitter(3.0))

        # =============================================================
        # CHECK IF IN MEETING
        # =============================================================
        in_meeting = False
        try:
            check = await page.evaluate("""() => {
                const body = (document.body?.innerText || '').toLowerCase();
                return body.includes('leave meeting') || body.includes('leave') || 
                       document.getElementById('wc-footer') !== null ||
                       document.querySelector('[class*="unmute"], [class*="mute"]') !== null;
            }""")
            in_meeting = check
        except:
            pass

        if in_meeting:
            print(f"[10] ✅ SUCCESS: IN THE MEETING!")
        else:
            print(f"[10] ❌ NOT in meeting. URL: {page.url}")

        await page.screenshot(path=f"10_final_{first_name}.png")

        # =============================================================
        # STAY IN MEETING
        # =============================================================
        print(f"[11] Staying for {duration//60} min...")
        start = time.time()
        while time.time() - start < duration:
            await asyncio.sleep(45)
            elapsed = int(time.time() - start)
            if elapsed % 300 < 10:
                print(f"    {elapsed//60}m {elapsed%60}s")
            if elapsed % 120 < 5:
                try:
                    ok = await page.evaluate("""() => (document.body?.innerText || '').toLowerCase().includes('leave')""")
                    if not ok:
                        break
                except:
                    break

        print(f"[12] Done ✓")
        await browser.close()


async def main():
    global JOIN_URL, BOT1_FIRST, BOT1_LAST, BOT1_EMAIL
    global BOT2_FIRST, BOT2_LAST, BOT2_EMAIL, DURATION_SECONDS

    if len(sys.argv) >= 2 and sys.argv[1] and sys.argv[1].startswith("http"):
        JOIN_URL = sys.argv[1]

    print("=" * 60)
    print("  ZOOM BOT — FAKE LIVE LIKES")
    print(f"  Bot 1: {BOT1_FIRST} {BOT1_LAST} <{BOT1_EMAIL}>")
    print(f"  Bot 2: {BOT2_FIRST} {BOT2_LAST} <{BOT2_EMAIL}>")
    print(f"  Duration: {DURATION_SECONDS//60} min each")
    print("=" * 60)

    # Bot 1
    display1 = f"{BOT1_FIRST} {BOT1_LAST}"
    await run_single_bot(BOT1_FIRST, BOT1_LAST, BOT1_EMAIL, display1, DURATION_SECONDS)

    # Bot 2
    display2 = f"{BOT2_FIRST} {BOT2_LAST}"
    await run_single_bot(BOT2_FIRST, BOT2_LAST, BOT2_EMAIL, display2, DURATION_SECONDS)

    print("\n✅ BOTH BOTS COMPLETED SUCCESSFULLY")


if __name__ == "__main__":
    asyncio.run(main())

File: zoomBot.js
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

// Configuration from environment variables
const MEETING_URL = process.env.MEETING_URL;
const BOT_NAME = process.env.BOT_NAME || 'Rahul Sharma';
const MEETING_PASSWORD = process.env.MEETING_PASSWORD || '';
const STAY_DURATION = parseInt(process.env.STAY_DURATION || '3600');
const SCREENSHOT_DIR = process.env.SCREENSHOT_DIR || 'screenshots';

// Ensure screenshot directory exists
if (!fs.existsSync(SCREENSHOT_DIR)) {
  fs.mkdirSync(SCREENSHOT_DIR, { recursive: true });
}

function log(msg) {
  const timestamp = new Date().toISOString();
  console.log(`[${timestamp}] [${BOT_NAME}] ${msg}`);
}

async function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function takeScreenshot(page, name) {
  const filePath = path.join(SCREENSHOT_DIR, `${BOT_NAME.replace(/\s+/g, '_')}_${name}.png`);
  await page.screenshot({ path: filePath, fullPage: true });
  log(`Screenshot saved: ${filePath}`);
  return filePath;
}

async function clickButton(page, buttonText, timeout = 15000, waitAfter = 5000) {
  try {
    // Try multiple selector strategies
    const button = page.locator(`button:has-text("${buttonText}")`);
    await button.waitFor({ state: 'visible', timeout });
    await button.scrollIntoViewIfNeeded();
    await button.click({ force: true });
    log(`Clicked button: "${buttonText}"`);
    await sleep(waitAfter);
    return true;
  } catch (e) {
    log(`Could not click "${buttonText}" via button selector: ${e.message}`);
    
    try {
      // Fallback: try span or div with that text
      const el = page.locator(`text="${buttonText}"`).first();
      await el.waitFor({ state: 'visible', timeout: 5000 });
      await el.click({ force: true });
      log(`Clicked text: "${buttonText}" (fallback)`);
      await sleep(waitAfter);
      return true;
    } catch (e2) {
      log(`Fallback also failed for "${buttonText}": ${e2.message}`);
      return false;
    }
  }
}

async function run() {
  const videoPath = path.join(SCREENSHOT_DIR, `${BOT_NAME.replace(/\s+/g, '_')}_recording.webm`);
  
  const browser = await chromium.launch({
    headless: true,
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
      '--disable-gpu',
      '--mute-audio',
      '--disable-web-security',
      '--disable-features=IsolateOrigins,site-per-process',
      '--no-first-run',
      '--no-default-browser-check',
      '--disable-blink-features=AutomationControlled',
      '--use-fake-ui-for-media-stream',
      '--use-fake-device-for-media-stream',
    ],
  });

  // Create context with video recording
  const context = await browser.newContext({
    viewport: { width: 1280, height: 720 },
    userAgent: 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36',
    locale: 'en-US',
    timezoneId: 'Asia/Kolkata',
    permissions: ['camera', 'microphone'],
    recordVideo: { dir: SCREENSHOT_DIR, size: { width: 1280, height: 720 } },
  });

  const page = await context.newPage();

  try {
    log(`Starting bot for meeting URL`);
    log(`Bot name: ${BOT_NAME}`);
    log(`Stay duration: ${STAY_DURATION}s`);

    // Step 0: Navigate to meeting URL
    log('Navigating to meeting URL...');
    await page.goto(MEETING_URL, { 
      waitUntil: 'networkidle',
      timeout: 60000 
    });
    await takeScreenshot(page, '00_landed');
    await sleep(5000);

    // Step 1: Click "Join from your Browser" / "Join from Browser"
    log('Step 1: Looking for "Join from Browser" link...');
    
    // Multiple possible selectors for the "Join from Browser" link
    const joinFromBrowserSelectors = [
      'a:has-text("Join from your Browser")',
      'a:has-text("Join from Browser")',
      'a:has-text("join from your browser")',
      'button:has-text("Join from your Browser")',
      'button:has-text("Join from Browser")',
      'text=Join from your Browser',
      'text=Join from Browser',
      '#zoom-ui-frame',
      'a[href*="wc/join"]',
      '.join-from-browser',
    ];
    
    let joinFromBrowserClicked = false;
    for (const selector of joinFromBrowserSelectors) {
      try {
        const el = page.locator(selector).first();
        if (await el.isVisible({ timeout: 3000 })) {
          await el.click({ force: true });
          log(`Clicked "Join from Browser" using selector: ${selector}`);
          joinFromBrowserClicked = true;
          break;
        }
      } catch (e) {
        // continue trying
      }
    }
    
    if (!joinFromBrowserClicked) {
      log('Could not find "Join from Browser" - may already be on meeting page');
    }
    
    await takeScreenshot(page, '01_after_join_from_browser');
    await sleep(5000);

    // Step 2: Handle mic/camera permission popup — "Continue without microphone and camera"
    log('Step 2: Handling mic/camera dialog...');
    
    // The permission dialog might be in an iframe or the main page
    // Check main page first
    let dialogHandled = await clickButton(page, 'Continue without microphone and camera', 5000, 5000);
    
    if (!dialogHandled) {
      // Check for any dialog-like element
      try {
        const dialogEls = ['div[role="dialog"]', '.permission-dialog', '[class*="permission"]', '[class*="dialog"]'];
        for (const sel of dialogEls) {
          const els = page.locator(sel);
          const count = await els.count();
          if (count > 0) {
            log(`Found ${count} dialog elements with selector: ${sel}`);
            // Try clicking any button inside
            const btn = els.first().locator('button, a, span').filter({ hasText: /continue|Allow|accept|ok/i }).first();
            if (await btn.isVisible({ timeout: 2000 })) {
              await btn.click({ force: true });
              log('Clicked button inside dialog');
              dialogHandled = true;
              break;
            }
          }
        }
      } catch (e) {
        log(`Error handling dialog: ${e.message}`);
      }
    }
    
    await takeScreenshot(page, '02_after_permission_dialog');
    await sleep(5000);

    // Step 3: Second "Continue without microphone and camera" (the actual join page)
    log('Step 3: Looking for second permission dialog...');
    dialogHandled = await clickButton(page, 'Continue without microphone and camera', 5000, 5000);
    
    // If the exact text doesn't match, try partial match
    if (!dialogHandled) {
      try {
        const btn = page.locator('button').filter({ hasText: /continue without/i }).first();
        if (await btn.isVisible({ timeout: 3000 })) {
          await btn.click({ force: true });
          log('Clicked button matching "continue without" (regex)');
          dialogHandled = true;
        }
      } catch (e) {
        log(`Continue without regex match failed: ${e.message}`);
      }
    }
    
    await takeScreenshot(page, '03_after_second_permission');
    await sleep(5000);

    // Step 4: Click "Join" button
    log('Step 4: Looking for Join button...');
    
    // Check if there's a name input that needs filling
    try {
      const nameInput = page.locator('input[placeholder*="Name"], input[name*="name"], input[id*="name"], input[placeholder*="name"]').first();
      if (await nameInput.isVisible({ timeout: 3000 })) {
        await nameInput.fill('');
        await nameInput.fill(BOT_NAME);
        log(`Filled name input with: ${BOT_NAME}`);
        await sleep(1000);
      }
    } catch (e) {
      log(`No name input found or error: ${e.message}`);
    }
    
    // Try to click Join
    let joinClicked = await clickButton(page, 'Join', 10000, 5000);
    
    if (!joinClicked) {
      // Try multiple Join button selectors
      const joinSelectors = [
        'button#joinBtn',
        'button.join-btn',
        'button[class*="join"]',
        'button:has-text("Join")',
        'a:has-text("Join")',
        'input[type="submit"][value*="Join"]',
        '#joinBtn',
        '[data-testid*="join"]',
        '.meeting-join-btn',
      ];
      
      for (const sel of joinSelectors) {
        try {
          const btn = page.locator(sel).first();
          if (await btn.isVisible({ timeout: 3000 })) {
            await btn.click({ force: true });
            log(`Clicked Join with selector: ${sel}`);
            joinClicked = true;
            break;
          }
        } catch (e) {
          // continue
        }
      }
    }
    
    if (!joinClicked) {
      log('WARNING: Could not find Join button - taking screenshot for debugging');
    }
    
    await takeScreenshot(page, '04_after_join_click');
    await sleep(10000);

    // Check if we successfully joined
    await takeScreenshot(page, '05_post_join_check');
    
    // Look for in-meeting indicators
    const inMeetingIndicators = [
      'button[aria-label*="Mute"]',
      'button[aria-label*="Leave"]',
      'button[aria-label*="Leave Meeting"]',
      '[class*="participant"]',
      '[class*="in-meeting"]',
      '[class*="meeting-ui"]',
      'text=Participants',
      'text=Leave',
    ];
    
    let inMeeting = false;
    for (const sel of inMeetingIndicators) {
      try {
        if (await page.locator(sel).first().isVisible({ timeout: 3000 })) {
          log(`SUCCESS: In-meeting indicator found: ${sel}`);
          inMeeting = true;
          break;
        }
      } catch (e) {}
    }
    
    if (!inMeeting) {
      log('WARNING: No in-meeting indicators found - may not have joined successfully');
      // Check if we're in a waiting room
      try {
        const waitingRoom = page.locator('text=Waiting for the host').first();
        if (await waitingRoom.isVisible({ timeout: 3000 })) {
          log('In waiting room - waiting for host to admit');
          inMeeting = true; // Treat waiting room as "in meeting"
        }
      } catch (e) {}
    }

    if (inMeeting) {
      log(`Bot successfully joined! Staying for ${STAY_DURATION} seconds...`);
      
      // Take periodic screenshots during the stay
      const screenshotInterval = 300; // every 5 minutes
      const totalScreenshots = Math.floor(STAY_DURATION / screenshotInterval);
      
      for (let i = 0; i < totalScreenshots; i++) {
        await sleep(screenshotInterval * 1000);
        await takeScreenshot(page, `06_staying_${i + 1}`);
        log(`Still in meeting... (${((i + 1) * screenshotInterval / 60).toFixed(0)}min elapsed)`);
      }
      
      log('Stay duration completed. Leaving meeting.');
    } else {
      log('Could not confirm meeting join - staying in browser for debugging');
      await sleep(30000); // extra wait for screenshots
    }

  } catch (error) {
    log(`ERROR: ${error.message}`);
    console.error(error.stack);
    await takeScreenshot(page, '99_error_state');
  } finally {
    await takeScreenshot(page, '99_final');
    await page.close();
    await context.close();
    await browser.close();
    log('Bot finished.');
  }
}

run().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});
